src/protocols/handshake.rs

```diff
@@ -46,38 +48,117 @@ pub enum State {
     Done,
 }
 
+#[deprecated(since="0.2.9", note="please use `HandshakeProtocol.builder()` instead")]
 #[derive(Debug, PartialEq)]
 pub enum ConnectionType {
     Tcp,
     Unix,
 }
 
+pub struct HandshakeBuilder {
+    role: Agency,
+    variant: Variant,
+    magic: u32,
+}
+
+#[derive(Clone, Copy)]
+enum Variant {
+    N2N,
+    C2N,
+}
+
+impl HandshakeBuilder {
+    pub fn network_magic(&mut self, magic: u32) -> &mut Self {
+        self.magic = magic;
+        self
+    }
+    pub fn client(&mut self) -> &mut Self {
+        self.role = Agency::Client;
+        self
+    }
+    pub fn server(&mut self) -> &mut Self {
+        self.role = Agency::Server;
+        self
+    }
+    pub fn client_to_node(&mut self) -> &mut Self {
+        self.variant = Variant::C2N;
+        self
+    }
+    pub fn node_to_node(&mut self) -> &mut Self {
+        self.variant = Variant::N2N;
+        self
+    }
+    pub fn build(&mut self) -> Result<HandshakeProtocol, Error> {
+        Ok(HandshakeProtocol {
+            role: self.role,
+            variant: self.variant,
+            network_magic: self.magic,
+            state: State::Propose,
+            result: None,
+        })
+    }
```

Comment:
why add an error if no error can be returned ?

Reply:
> why add an error if no error can be returned ?

That is in order to be future-proof. Builders can be extended without introducing a breaking change for consumer code. There might be validation needed in any of the builders in the future.

Reply:
@NicolasDP do you consider this satisfactory answer, if so I am ready to merge it in...